# Create ISA-API Investigation from Datascriptor Study Design configuration
# Crossover Study with radiological and drug treatment on Mice

In this notebook I will show you how you can use a study design configuration is JSON format as produce by datascriptor (https://gitlab.com/datascriptor/datascriptor) to generate a single-study ISA investigation and how you can then serialise it in JSON and tabular (i.e. CSV) format.

Or study design configuration consists of:
- a 4-arm study design. Each arm has 10 subjects
- subjects are mice. There is an observational factor, named "sex" with two values: "M" and "F"
- a crossover of two treatments, a drug treatment (cyclophosphamide 500 mg for 20 days) and a radiological treatment ("strontium chloride sr 89" 2 MBq/Kg forr 40 days)
- four non-treatment phases: screen, run-in, washout and follow-up
- three sample types colllected: tissue, blood, and urine
- three assay types: 
    - metabolite profiling through mass spectrometry on the urine sample.
    - clinical chemistry with marker on blood samples
    - genomic sequencing on tissue samples

## 1. Setup

Let's import all the required libraries

In [1]:
from time import time
import os
import json

In [2]:
## ISA-API related imports
from isatools.model import Investigation, Study

In [3]:
## ISA-API create mode related imports
from isatools.create.model import StudyDesign
from isatools.create.connectors import generate_study_design

# serializer from ISA Investigation to JSON
from isatools.isajson import ISAJSONEncoder

# ISA-Tab serialisation
from isatools import isatab

In [4]:
## ISA-API create mode related imports
from isatools.create import model
from isatools import isajson

## 2. Load the Study Design JSON configuration

First of all we load the study design configurator with all the specs defined above

In [5]:
with open(os.path.abspath(os.path.join(
    "config", "crossover-study-design-2-treatments-mice.json"
)), "r") as config_file:
    study_design_config = json.load(config_file)
study_design_config

{'name': 'Test study',
 'description': 'This design describes a study with repeated treatments on the same arms. We will be collecting three samples and these will undergo three different assays.',
 'subjectType': 'Mouse',
 'subjectSize': 10,
 'designType': {'term': 'crossover design',
  'id': 'OBI:0500003',
  'iri': 'http://purl.obolibrary.org/obo/OBI_0500003',
  'label': 'Study subjects receive repeated treatments',
  'value': 'crossover'},
 'observationalFactors': [{'name': 'sex',
   'values': ['M', 'F'],
   'isQuantitative': False,
   'unit': None}],
 'subjectGroups': {'selected': [{'name': 'SubjectGroup_0',
    'type': 'Mouse',
    'characteristics': [{'name': 'sex',
      'value': 'M',
      'unit': None,
      'isQuantitative': False}]},
   {'name': 'SubjectGroup_1',
    'type': 'Mouse',
    'characteristics': [{'name': 'sex',
      'value': 'F',
      'unit': None,
      'isQuantitative': False}]}],
  'unselected': []},
 'treatmentPlan': {'screen': {'selected': True,
   'name':

## 3. Generate the ISA Study Design from the JSON configuration
To perform the conversion we just need to use the function `generate_isa_study_design()` (name possibly subject to change, should we drop the "isa" and "datascriptor" qualifiers?)

In [6]:
study_design = generate_study_design(study_design_config)
assert isinstance(study_design, StudyDesign)

## 4. Generate the ISA Study from the StudyDesign and embed it into an ISA Investigation

The `StudyDesign.generate_isa_study()` method returns the complete ISA-API `Study` object.

In [7]:
start = time()
study = study_design.generate_isa_study()
end = time()
print('The generation of the study design took {:.2f} s.'.format(end - start))
assert isinstance(study, Study)
investigation = Investigation(identifier='inv01', studies=[study])

The generation of the study design took 1.50 s.


## 5. Serialize and save the JSON representation of the generated ISA Investigation

In [8]:
start = time()
inv_json = json.dumps(investigation, cls=ISAJSONEncoder, sort_keys=True, indent=4, separators=(',', ': '))
end = time()
print('The JSON serialisation of the ISA investigation took {:.2f} s.'.format(end - start))

The JSON serialisation of the ISA investigation took 0.67 s.


In [9]:
directory = os.path.abspath(os.path.join('output', 'crossover-2-treatments-mice'))
os.makedirs(directory, exist_ok=True)
with open(os.path.abspath(os.path.join(directory, 'isa-investigation-crossover-2-treatments-mice.json')), 'w') as out_fp:
    json.dump(json.loads(inv_json), out_fp)

## 6. Dump the ISA Investigation to ISA-Tab

In [10]:
start = time()
isatab.dump(investigation, directory)
end = time()
print('The Tab serialisation of the ISA investigation took {:.2f} s.'.format(end - start))

The Tab serialisation of the ISA investigation took 32.09 s.


To use them on the notebook we can also dump the tables to pandas DataFrames, using the `dump_tables_to_dataframes` function rather than dump

In [11]:
dataframes = isatab.dump_tables_to_dataframes(investigation)

In [12]:
len(dataframes)

4

## 7. Check the correctness of the ISA-Tab DataFrames 

We have 1 study file and 2 assay files (one for MS and one for NMR). Let's check the names:

In [13]:
for key in dataframes.keys():
    display(key)

's_study_01.txt'

'a_AT8_genome-sequencing_nucleic-acid-sequencing.txt'

'a_AT11_clinical-chemistry_marker-panel.txt'

'a_AT1_metabolite-profiling_mass-spectrometry.txt'

### 7.1 Count of subjects and samples

We have 10 subjects in the each of the six arms for a total of 60 subjects. 5 blood samples per subject are collected (1 in treatment 1 phase, 1 in treatment, and 3 in the follow-up phase) for a total of 300 blood samples. These will undergo the NMR assay. We have 4 saliva samples per subject (1 during screen and 3 during follow-up) for a total of 240 saliva samples. These will undergo the "mass spcetrometry" assay.

In [14]:
study_frame = dataframes['s_study_01.txt']
count_arm0_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP0' in el)])
count_arm1_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP1' in el)])
count_arm2_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP2' in el)])
count_arm3_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP3' in el)])
print("There are {} samples in the GRP0 arm (i.e. group)".format(count_arm0_samples))
print("There are {} samples in the GRP1 arm (i.e. group)".format(count_arm1_samples))
print("There are {} samples in the GRP2 arm (i.e. group)".format(count_arm2_samples))
print("There are {} samples in the GRP3 arm (i.e. group)".format(count_arm3_samples))

There are 160 samples in the GRP0 arm (i.e. group)
There are 160 samples in the GRP1 arm (i.e. group)
There are 160 samples in the GRP2 arm (i.e. group)
There are 160 samples in the GRP3 arm (i.e. group)


In [15]:
study_frame

Source Name Characteristics[Study Subject] Characteristics[sex]  \
0    GRP0_SBJ01                          Mouse                    M   
1    GRP0_SBJ01                          Mouse                    M   
2    GRP0_SBJ01                          Mouse                    M   
3    GRP0_SBJ01                          Mouse                    M   
4    GRP0_SBJ01                          Mouse                    M   
..          ...                            ...                  ...   
635  GRP3_SBJ10                          Mouse                    F   
636  GRP3_SBJ10                          Mouse                    F   
637  GRP3_SBJ10                          Mouse                    F   
638  GRP3_SBJ10                          Mouse                    F   
639  GRP3_SBJ10                          Mouse                    F   

          Protocol REF Parameter Value[Sampling order]  \
0    sample collection                             033   
1    sample collection                             056   
2    sample collection                             015   
3    sample collection                             117   
4    sample collection                             118   
..                 ...                             ...   
635  sample collection                             582   
636  sample collection                             516   
637  sample collection                             541   
638  sample collection                             502   
639  sample collection                             581   

    Parameter Value[Study cell]        Date Performer  \
0                          A0E1  2021-04-14   Unknown   
1                          A0E2  2021-04-14   Unknown   
2                          A0E0  2021-04-14   Unknown   
3                          A0E5  2021-04-14   Unknown   
4                          A0E5  2021-04-14   Unknown   
..                          ...         ...       ...   
635                        A3E4  2021-04-14   Unknown   
636                        A3E1  2021-04-14   Unknown   
637                        A3E2  2021-04-14   Unknown   
638                        A3E0  2021-04-14   Unknown   
639                        A3E4  2021-04-14   Unknown   

                            Sample Name Characteristics[organism part]  \
0           GRP0_SBJ01_A0E1_SMP-Urine-1                          Urine   
1    GRP0_SBJ01_A0E2_SMP-Blood-Sample-2                   Blood Sample   
2           GRP0_SBJ01_A0E0_SMP-Urine-1                          Urine   
3    GRP0_SBJ01_A0E5_SMP-Blood-Sample-1                   Blood Sample   
4    GRP0_SBJ01_A0E5_SMP-Blood-Sample-2                   Blood Sample   
..                                  ...                            ...   
635  GRP3_SBJ10_A3E4_SMP-Blood-Sample-2                   Blood Sample   
636         GRP3_SBJ10_A3E1_SMP-Urine-1                          Urine   
637  GRP3_SBJ10_A3E2_SMP-Blood-Sample-1                   Blood Sample   
638         GRP3_SBJ10_A3E0_SMP-Urine-2                          Urine   
639  GRP3_SBJ10_A3E4_SMP-Blood-Sample-1                   Blood Sample   

    Comment[study step with treatment] Factor Value[Sequence Order]  \
0                                   NO                            1   
1                                  YES                            2   
2                                   NO                            0   
3                                   NO                            5   
4                                   NO                            5   
..                                 ...                          ...   
635                                YES                            4   
636                                 NO                            1   
637                                YES                            2   
638                                 NO                            0   
639                                YES                            4   

          Factor Value[AGEN

In [16]:
dataframes['a_AT1_metabolite-profiling_mass-spectrometry.txt']

Sample Name Comment[study step with treatment]  \
0     GRP0_SBJ01_A0E0_SMP-Urine-1                                 NO   
1     GRP0_SBJ01_A0E0_SMP-Urine-1                                 NO   
2     GRP0_SBJ01_A0E0_SMP-Urine-1                                 NO   
3     GRP0_SBJ01_A0E0_SMP-Urine-1                                 NO   
4     GRP0_SBJ01_A0E0_SMP-Urine-2                                 NO   
...                           ...                                ...   
1115  GRP3_SBJ10_A3E5_SMP-Urine-1                                 NO   
1116  GRP3_SBJ10_A3E5_SMP-Urine-2                                 NO   
1117  GRP3_SBJ10_A3E5_SMP-Urine-2                                 NO   
1118  GRP3_SBJ10_A3E5_SMP-Urine-2                                 NO   
1119  GRP3_SBJ10_A3E5_SMP-Urine-2                                 NO   

     Protocol REF Performer         Extract Name  \
0      extraction   Unknown    AT1-S5-Extract-R1   
1      extraction   Unknown    AT1-S5-Extract-R1   
2      extraction   Unknown    AT1-S5-Extract-R1   
3      extraction   Unknown    AT1-S5-Extract-R1   
4      extraction   Unknown    AT1-S6-Extract-R1   
...           ...       ...                  ...   
1115   extraction   Unknown  AT1-S271-Extract-R1   
1116   extraction   Unknown  AT1-S272-Extract-R1   
1117   extraction   Unknown  AT1-S272-Extract-R1   
1118   extraction   Unknown  AT1-S272-Extract-R1   
1119   extraction   Unknown  AT1-S272-Extract-R1   

     Characteristics[extract type] Term Accession Number Protocol REF.1  \
0                   polar fraction        polar fraction       labeling   
1                   polar fraction        polar fraction       labeling   
2                   polar fraction        polar fraction       labeling   
3                   polar fraction        polar fraction       labeling   
4                   polar fraction        polar fraction       labeling   
...                            ...                   ...            ...   
1115                polar fraction        polar fraction       labeling   
1116                polar fraction        polar fraction       labeling   
1117                polar fraction        polar fraction       labeling   
1118                polar fraction        polar fraction       labeling   
1119                polar fraction        polar fraction       labeling   

     Performer.1 Labeled Extract Name  ...     Protocol REF.2  \
0        Unknown         AT1-S5-LE-R1  ...  mass spectrometry   
1        Unknown         AT1-S5-LE-R1  ...  mass spectrometry   
2        Unknown         AT1-S5-LE-R1  ...  mass spectrometry   
3        Unknown         AT1-S5-LE-R1  ...  mass spectrometry   
4        Unknown         AT1-S6-LE-R1  ...  mass spectrometry   
...          ...                  ...  ...                ...   
1115     Unknown       AT1-S271-LE-R1  ...  mass spectrometry   
1116     Unknown       AT1-S272-LE-R1  ...  mass spectrometry   
1117     Unknown       AT1-S272-LE-R1  ...  mass spectrometry   
1118     Unknown       AT1-S272-LE-R1  ...  mass spectrometry   
1119     Unknown       AT1-S272-LE-R1  ...  mass spectrometry   

     Parameter Value[instrument]                    Term Accession Number.1  \
0              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   
1              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   
2              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   
3              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   
4              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   
...                          ...                                        ...   
1115           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   
1116           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   
1117           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   
1118           Agilent QTQF 6510  http://purl.obolibrary.org

In [17]:
dataframes['a_AT11_clinical-chemistry_marker-panel.txt']

Sample Name Comment[study step with treatment]  \
0    GRP0_SBJ01_A0E0_SMP-Blood-Sample-1                                 NO   
1    GRP0_SBJ01_A0E0_SMP-Blood-Sample-1                                 NO   
2    GRP0_SBJ01_A0E0_SMP-Blood-Sample-1                                 NO   
3    GRP0_SBJ01_A0E2_SMP-Blood-Sample-1                                YES   
4    GRP0_SBJ01_A0E2_SMP-Blood-Sample-1                                YES   
..                                  ...                                ...   
955  GRP3_SBJ10_A3E5_SMP-Blood-Sample-2                                 NO   
956  GRP3_SBJ10_A3E5_SMP-Blood-Sample-2                                 NO   
957  GRP3_SBJ10_A3E5_SMP-Blood-Sample-3                                 NO   
958  GRP3_SBJ10_A3E5_SMP-Blood-Sample-3                                 NO   
959  GRP3_SBJ10_A3E5_SMP-Blood-Sample-3                                 NO   

           Protocol REF Performer               Raw Data File  
0    sample preparation   Unknown    AT11-S9-raw_data_file-R1  
1    sample preparation   Unknown    AT11-S7-raw_data_file-R1  
2    sample preparation   Unknown    AT11-S8-raw_data_file-R1  
3    sample preparation   Unknown   AT11-S43-raw_data_file-R1  
4    sample preparation   Unknown   AT11-S45-raw_data_file-R1  
..                  ...       ...                         ...  
955  sample preparation   Unknown  AT11-S921-raw_data_file-R1  
956  sample preparation   Unknown  AT11-S919-raw_data_file-R1  
957  sample preparation   Unknown  AT11-S923-raw_data_file-R1  
958  sample preparation   Unknown  AT11-S924-raw_data_file-R1  
959  sample preparation   Unknown  AT11-S922-raw_data_file-R1  

[960 rows x 5 columns]

In [18]:
dataframes['a_AT8_genome-sequencing_nucleic-acid-sequencing.txt']

Sample Name Comment[study step with treatment]  \
0    GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1                                YES   
1    GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1                                YES   
2    GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1                                YES   
3    GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1                                YES   
4    GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1                                YES   
..                                   ...                                ...   
635  GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1                                YES   
636  GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1                                YES   
637  GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1                                YES   
638  GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1                                YES   
639  GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1                                YES   

    Protocol REF Performer        Extract Name Characteristics[extract type]  \
0     extraction   Unknown   AT8-S6-Extract-R1                          gDNA   
1     extraction   Unknown   AT8-S5-Extract-R2                           DNA   
2     extraction   Unknown   AT8-S6-Extract-R1                          gDNA   
3     extraction   Unknown   AT8-S6-Extract-R2                           DNA   
4     extraction   Unknown   AT8-S6-Extract-R1                          gDNA   
..           ...       ...                 ...                           ...   
635   extraction   Unknown  AT8-S71-Extract-R2                           DNA   
636   extraction   Unknown  AT8-S71-Extract-R2                           DNA   
637   extraction   Unknown  AT8-S72-Extract-R2                           DNA   
638   extraction   Unknown  AT8-S72-Extract-R1                          gDNA   
639   extraction   Unknown  AT8-S72-Extract-R1                          gDNA   

          Protocol REF.1 Parameter Value[instrument]  \
0    library_preparation       Illumina NovaSeq 6000   
1    library_preparation       Illumina NovaSeq 6000   
2    library_preparation       Illumina NovaSeq 6000   
3    library_preparation       Illumina NovaSeq 6000   
4    library_preparation       Illumina NovaSeq 6000   
..                   ...                         ...   
635  library_preparation       Illumina NovaSeq 6000   
636  library_preparation       Illumina NovaSeq 6000   
637  library_preparation       Illumina NovaSeq 6000   
638  library_preparation       Illumina NovaSeq 6000   
639  library_preparation       Illumina NovaSeq 6000   

    Parameter Value[library_orientation] Parameter Value[library_strategy]  \
0                                 single                               WGS   
1                                 single                               WGS   
2                                 single                          AMPLICON   
3                                 single                          AMPLICON   
4                                 single                               WGS   
..                                   ...                               ...   
635                               single                          AMPLICON   
636                               single                               WGS   
637                               single                          AMPLICON   
638                               single                               WGS   
639                               single                          AMPLICON   

    Performer.1                 Raw Data File  
0       Unknown   AT8-S6-raw_data_file-R1.raw  
1       Unknown   AT8-S5-raw_data_file-R6.raw  
2       Unknown   AT8-S6-raw_data_file-R3.raw  
3       Unknown   AT8-S6-raw_data_file-R8.raw  
4       Unknown   AT8-S6-raw_data_file-R2.raw  
..          ...                           ...  
635     Unknown  AT8-S71-raw_data_file-R8.raw  
636     Unknown  AT8-S71-raw_data_file-R5.raw  
637     Unknown  AT8-S72-raw_data_file-R8.raw  
638     Unknown  AT8-S72-raw_

# 7.1 Overview of the Mass Spec assay table

For the mass. spec. assay table, we have 160 (derma) samples, 160 extracts (1 per  sample, "polar" fraction), 320 labeled extracts (2 per extract, as "#replicates" is  2) and 640 mass spec protocols + 640 output files (2 per labeled extract as we do 1 technical replicate with 2 protocol parameter combinations `["Agilent QTQF 6510", "FIA", "positive mode"]` and `["Agilent QTQF 6510", "LC", "positive mode"]`).

In [19]:
dataframes['a_AT1_metabolite-profiling_mass-spectrometry.txt'].nunique(axis=0, dropna=True)

Sample Name                            280
Comment[study step with treatment]       2
Protocol REF                             1
Performer                                1
Extract Name                           280
Characteristics[extract type]            1
Term Accession Number                    1
Protocol REF.1                           1
Performer.1                              1
Labeled Extract Name                   280
Label                                    1
Protocol REF.2                           1
Parameter Value[instrument]              1
Term Accession Number.1                  1
Parameter Value[injection_mode]          2
Term Accession Number.2                  2
Parameter Value[acquisition_mode]        1
Term Accession Number.3                  1
MS Assay Name                         1120
Performer.2                              1
Raw Spectral Data File                1120
dtype: int64

### Overview of the Genome Sequencing assay table

For the NMR assay table, we have 360 (blood) samples, 720 extracts (2 per  sample, a single replicate of the "supernatant" and "pellet" fractions) and 2880 NMR protocols + 4800 output files (4 per extract as we do 2 technical replicates with 2 protocol parameter combinations `["Bruker Avance II 1 GHz", "1D 1H NMR", "CPGM"]` and `["Bruker Avance II 1 GHz", "1D 1H NMR", "TOCSY"]`).

In [20]:
dataframes['a_AT8_genome-sequencing_nucleic-acid-sequencing.txt'].nunique(axis=0, dropna=True)

Sample Name                              40
Comment[study step with treatment]        1
Protocol REF                              1
Performer                                 1
Extract Name                            160
Characteristics[extract type]             2
Protocol REF.1                            1
Parameter Value[instrument]               1
Parameter Value[library_orientation]      1
Parameter Value[library_strategy]         2
Performer.1                               1
Raw Data File                           640
dtype: int64